keps/sig-node/4010-head-request-probes/README.md

```diff
@@ -0,0 +1,265 @@
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [X] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This Kubernetes Enhancement Proposal (KEP) introduces the capability to use HTTP HEAD requests for defining readiness and liveness probes in Kubernetes pods. Currently, only GET requests are supported for probes, limiting the options for checking the health of containers. By allowing the use of HEAD requests, Kubernetes users will have more flexibility and finer-grained control over their probes, especially for customer-facing applications.
+
+## Motivation
+
+The motivation behind this proposal is to enhance Kubernetes' readiness and liveness probes by supporting HTTP HEAD requests. While GET requests are commonly used, they might not always be suitable for certain scenarios, such as checking the availability of resources without fetching their content. The ability to use HEAD requests will enable more efficient health checks for specific use cases, particularly those involving customer-facing services. By extending the probe functionality, Kubernetes aims to provide a more versatile and customizable platform for managing container health.
```

Comment:
I'd really want to see some specific cases where GET was untenable, but HEAD would have been OK, where there wasn't something else that could have been done differently.  I'm not saying it doesn't exist, but every API change needs to be justified.

What I mean is that there's a trivial workaround - run another container as a "probe proxy" which receives a GET and does a HEAD to the real backend.  This is about 10 lines of code.
 
It's not the most elegant approach, but it works.  Can we show even 3 or 4 cases where it was not viable?  Or heck, just show me a few real examples where this was necessary, which will at least make the argument that this isn't a solution to a problem nobody has.

Reply:
#### Without HEAD

- Increases the cognitive load for developers as it is counter intuitive, that Kubernetes does support GET, but not HEAD
- Inefficient and waste of energy. In cases, where GET request return data in the body
- Not supporting HEAD can have cascading effects as apps, tests and documentation need to adjust to the limitations of Kubernetes
- According to the HTTP specification, HEAD would be the ideal candidate for probes

#### Explanation

I am a user of Kubernetes and I have used it now for several years. From my point of view the world looks a bit different.
As a user, I want to write apps which are maintainable and efficient. From my perspective Kubernetes should support me in my work. I don't want to create workarounds to get my work done.

HTTP is a standardized protocol. It uses methods such as GET, HEAD, POST etc. The intention of HEAD is to do the same job as GET, but: "**MUST NOT** send content in the response" [RFC 9110 - 9.3.2. HEAD](https://datatracker.ietf.org/doc/html/rfc9110#section-9.3.2). I would argue, that HEAD is the ideal HTTP method for Kubernetes probes, based on the RFC 9110 definition.

At [SecHub](https://github.com/mercedes-benz/sechub/) we were doing exactly that, we implemented a health check using HEAD. We did not and do not develop SecHub to **only** run on Kubernetes. Once, we started deploying SecHub to Kubernetes we discovered, that Kubernetes does not support probes via HEAD, therefore we added a [GET request](https://github.com/mercedes-benz/sechub/issues/378). The task itself was easy, but the consequences are annoying. Now, we had to write [tests](https://github.com/mercedes-benz/sechub/blob/develop/sechub-doc/src/test/java/com/mercedesbenz/sechub/restdoc/AnonymousCheckAliveRestDocTest.java) and [document](https://mercedes-benz.github.io/sechub/latest/sechub-restapi.html#section-ucrestdoc-uc-UC_039) both: HEAD and GET health checks. We need to account for that in our [OpenAPI definition](https://github.com/mercedes-benz/sechub/releases/download/v0.43.0-server/sechub-openapi3-0.43.0.json) and in our client library which is based on the OpenAPI definition. Basically, doubling the work for a simple endpoint.

Next, we introduced another app [SecHub PDS](https://github.com/mercedes-benz/sechub/issues/918). Same issue, we implemented HEAD, next we had to add the [GET method](https://github.com/mercedes-benz/sechub/issues/918).

Furthermore, it is counter intuitive to have GET available for probes, but not HEAD. Especially, because Kubernetes only cares about the HTTP response codes of the probe requests and not the body. It increases the cognitive load. First, one has to understand the HTTP protocol next one has to learn about the limitations of Kubernetes regarding possible HTTP methods in probes. I, for my part, much rather prefer to learn about the HTTP protocol as it helps me outside of Kubernetes.

Another issue is energy efficiency (green computing/ICT sustainability). In cases, where the health check returns more than just an empty body for GET requests, being able to use a HEAD request can limit the resource consumption. Without a HEAD request the body is send and received to be discarded. Truly wasteful.
Those bytes need to be processed and send around, which cost CPU cycles and requires memory. Probes, especially livnessProbes, are getting executed several thousand times per day and million times in a year (depending on the configuration). All that computational effort adds up. At the end it is just wasted energy for nothing.

Examples of applications which implement a health check with HEAD, but return the full HTML page are:

- https://try.gitea.io/
        
        # GET request
        curl https://try.gitea.io/ -> returns the login page

        # HEAD request
        curl --head https://try.gitea.io/
        HTTP/1.1 200 Connection established

        HTTP/2 200 
        date: Wed, 07 Jun 2023 18:01:29 GMT
        server: Caddy

- https://forum.gitlab.com/ (running [Discourse](https://www.discourse.org/))

        # GET request
        curl https://forum.gitlab.com/

        # HEAD request
        curl --head https://forum.gitlab.com/
        HTTP/1.1 200 Connection established

        HTTP/2 200 
        server: nginx
        date: Wed, 07 Jun 2023 19:06:49 GMT
        content-type: text/html; charset=utf-8
        vary: Accept-Encoding

Remarks:

- Not everybody is lucky enough to be able to talk openly about their application and provide you with real examples. Working on a FOSS project provides me that freedom.
- I am not the original author of the issue: [Support httpHead for pod livenessProbe #49937](https://github.com/kubernetes/kubernetes/issues/49937), there are more users with a similar opinions

Reply:
I appreciate the energy you put into this response, truly.  

We must also acknowledge that Kube's probe definitions are used by other systems such as clouds ( https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-healthchecks.html, https://learn.microsoft.com/en-us/azure/load-balancer/load-balancer-custom-probe-overview, https://cloud.google.com/run/docs/configuring/healthchecks, just to pick a few) and none of those support HEAD.  Others like nginx also seem to only support GET, but I didn't find definitive docs.

ON the other hand, fastly supports HEAD (https://docs.fastly.com/en/guides/working-with-health-checks) as does haproxy (https://www.haproxy.com/documentation/hapee/latest/service-reliability/health-checking/active-checks/check-HTTP-service/)

I'm afraid that it's not as wide-spread as you might think.  It's not that I don't want to make things easier, but none of this stuff is cost-free and in nearly 10 years, I have heard this request only a handful of times.  I added some thoughts on some of @sftim apt comments, too.

Reply:
When I looked at cloud probes and saw the inconsistent `HEAD` support the counter argument for me was that some of them also do not support headers, while Kuberneetes does. Which makes this topic harder to reason about =)

I wanted to expand on the cost of supporting http HEAD. I don't worry about the engineering price that much. What bothers me is the supportability of HEAD probes. I saw a few issues when analyzing the response body of GET probe helped to find the problem with the probe - it was short circuited from some java servlet filter instead of reaching the actual GET routing. It is good when it short circuited to return error. But sometimes it will simply return 200 unconditionally rendering the probe useless. With HEAD - there is no body to analyze and troubleshooting will be hard. And I can see how HEAD may be misconfigured on different layers of the request routing leading to unexpected behaviors and support requests.

@Jeeppler what is your experience with HEAD probes supportability since you are using them? What do you use to confirm that HEAD probe ACTUALLY hits the callback that checks the app state?